packages/wrangler/src/deployment-bundle/bindings.ts

```diff
@@ -62,3 +69,258 @@ export function getBindings(
+		if (!d1.database_id) {
+			if (inBindingSettings(settings, "d1", d1.binding)) {
+				d1.database_id = INHERIT_SYMBOL;
+			} else {
+				pendingResources.d1_databases?.push({
+					binding: d1.binding,
+					async create(name) {
+						const db = await createD1Database(accountId, name);
+						d1.database_id = db.uuid;
+						return db.uuid;
+					},
+					updateId(id) {
+						d1.database_id = id;
+					},
+				});
+			}
+		}
+	}
+
+	if (Object.values(pendingResources).some((v) => v && v.length > 0)) {
+		logger.log();
+		printBindings(pendingResources, { provisioning: true });
+		logger.log();
+		if (pendingResources.kv_namespaces?.length) {
+			const preExistingKV = await listKVNamespaces(accountId, true);
+			await runProvisioningFlow(
+				pendingResources.kv_namespaces,
+				preExistingKV.map((ns) => ({ name: ns.title, id: ns.id })),
+				"KV Namespace",
+				"title or id",
+				scriptName,
+				autoCreate
+			);
+		}
+
+		if (pendingResources.d1_databases?.length) {
+			const preExisting = await listDatabases(accountId, true);
+			await runProvisioningFlow(
+				pendingResources.d1_databases,
+				preExisting.map((db) => ({ name: db.name, id: db.uuid })),
+				"D1 Database",
+				"name or id",
+				scriptName,
+				autoCreate
+			);
+		}
+		if (pendingResources.r2_buckets?.length) {
+			const preExisting = await listR2Buckets(accountId);
+			await runProvisioningFlow(
+				pendingResources.r2_buckets,
+				preExisting.map((bucket) => ({ name: bucket.name, id: bucket.name })),
+				"R2 Bucket",
+				"name",
+				scriptName,
+				autoCreate
+			);
+		}
+		logger.log(`🎉 All resources provisioned, continuing with deployment...\n`);
+	}
+}
+
+/** checks whether the binding id can be inherited from a prev deployment */
+function inBindingSettings<Type extends WorkerMetadataBinding["type"]>(
+	settings: Settings | undefined,
+	type: Type,
+	bindingName: string
+): Extract<WorkerMetadataBinding, { type: Type }> | undefined {
+	return settings?.bindings.find(
+		(binding): binding is Extract<WorkerMetadataBinding, { type: Type }> =>
+			binding.type === type && binding.name === bindingName
+	);
+}
+
+function getSettings(accountId: string, scriptName: string) {
+	return fetchResult<Settings>(
+		`/accounts/${accountId}/workers/scripts/${scriptName}/settings`
+	);
+}
+
+function printDivider() {
+	logger.log();
+	logger.log(chalk.dim("--------------------------------------"));
+	logger.log();
+}
+
+type NormalisedResourceInfo = {
+	name: string;
+	id: string;
+};
+type ResourceType = "d1_databases" | "r2_buckets" | "kv_namespaces";
+async function runProvisioningFlow(
+	pending: PendingResources[ResourceType],
+	preExisting: NormalisedResourceInfo[],
+	friendlyBindingName: string,
+	resourceKeyDescriptor: string,
+	scriptName: string,
+	autoCreate: boolean
+) {
+	const MAX_OPTIONS = 4;
+	if (pending.length) {
+		const options = preExisting
+			.map((resource) => ({
+				title: resource.name,
+				value: resource.id,
```

Comment:
Given we're mapping to `value`, why not just have `NormalisedResourceInfo` be `{ title: string, value: string }`?